File: mahjong_utils/models/hora_hand_pattern.py
from abc import ABC, abstractmethod
from typing import Optional, FrozenSet, Tuple

from pydantic import Field
from stringcase import pascalcase, snakecase

from mahjong_utils.models.hand_pattern import RegularHandPattern, HandPattern, ChitoiHandPattern, KokushiHandPattern
from mahjong_utils.models.tatsu import Tatsu
from mahjong_utils.models.tile import Tile, all_yaochu
from mahjong_utils.models.wind import Wind


class HoraHandPattern(HandPattern, ABC):
    agari: Tile
    tsumo: bool
    hu: int
    self_wind: Optional[Wind]
    round_wind: Optional[Wind]

    @abstractmethod
    def __encode__(self) -> dict:
        raise NotImplementedError()

    @classmethod
    def __decode__(cls, data: dict) -> "HoraHandPattern":
        if data['type'] == 'RegularHoraHandPattern':
            return RegularHoraHandPattern.__decode__(data)
        elif data['type'] == 'ChitoiHoraHandPattern':
            return ChitoiHoraHandPattern.__decode__(data)
        elif data['type'] == 'KokushiHoraHandPattern':
            return KokushiHoraHandPattern.__decode__(data)
        else:
            raise ValueError("invalid type: " + data['type'])


class RegularHoraHandPattern(HoraHandPattern, RegularHandPattern):
    agari_tatsu: Optional[Tatsu]
    jyantou: Tile

    def __encode__(self) -> dict:
        return dict(
            type="RegularHoraHandPattern",
            agari=self.agari.__encode__(),
            tsumo=self.tsumo,
            self_wind=pascalcase(self.self_wind.name) if self.self_wind is not None else None,
            round_wind=pascalcase(self.round_wind.name) if self.round_wind is not None else None,
            agari_tatsu=self.agari_tatsu.__encode__() if self.agari_tatsu is not None else None,
            pattern=RegularHandPattern.__encode__(self)
        )

    @classmethod
    def __decode__(cls, data: dict) -> "RegularHoraHandPattern":
        return RegularHoraHandPattern(
            agari=Tile.__decode__(data["agari"]),
            tsumo=data["tsumo"],
            self_wind=Wind[snakecase(data["selfWind"])] if data["selfWind"] is not None else None,
            round_wind=Wind[snakecase(data["roundWind"])] if data["roundWind"] is not None else None,
            agari_tatsu=Tatsu.__decode__(data["agariTatsu"]) if data["agariTatsu"] is not None else None,
            hu=data["hu"],
            **RegularHandPattern.__decode__(data["pattern"]).dict()
        )


class ChitoiHoraHandPattern(HoraHandPattern, ChitoiHandPattern):
    hu: int = 25
    remaining: Tuple[Tile, ...] = Field(default_factory=tuple)

    def __encode__(self) -> dict:
        return dict(
            type="ChitoiHoraHandPattern",
            pairs=[t.__encode__() for t in self.pairs],
            agari=self.agari.__encode__(),
            tsumo=self.tsumo,
            self_wind=pascalcase(self.self_wind.name) if self.self_wind is not None else None,
            round_wind=pascalcase(self.round_wind.name) if self.round_wind is not None else None,
        )

    @classmethod
    def __decode__(cls, data: dict) -> "ChitoiHoraHandPattern":
        return ChitoiHoraHandPattern(
            agari=Tile.__decode__(data["agari"]),
            pairs=frozenset(Tile.__decode__(t) for t in data["pairs"]),
            tsumo=data["tsumo"],
            self_wind=Wind[snakecase(data["selfWind"])] if data["selfWind"] is not None else None,
            round_wind=Wind[snakecase(data["roundWind"])] if data["roundWind"] is not None else None,
        )


class KokushiHoraHandPattern(HoraHandPattern, KokushiHandPattern):
    hu: int = 20
    repeated: Tile
    yaochu: FrozenSet[Tile] = all_yaochu
    remaining: Tuple[Tile, ...] = Field(default_factory=tuple)

    @property
    def thirteen_waiting(self) -> bool:
        return self.agari == self.repeated

    def __encode__(self) -> dict:
        return dict(
            type="KokushiHoraHandPattern",
            repeated=self.repeated.__encode__(),
            agari=self.agari.__encode__(),
            tsumo=self.tsumo,
            self_wind=pascalcase(self.self_wind.name) if self.self_wind is not None else None,
            round_wind=pascalcase(self.round_wind.name) if self.round_wind is not None else None,
        )

    @classmethod
    def __decode__(cls, data: dict) -> "KokushiHoraHandPattern":
        return KokushiHoraHandPattern(
            repeated=Tile.__decode__(data["repeated"]),
            agari=Tile.__decode__(data["agari"]),
            tsumo=data["tsumo"],
            self_wind=Wind[snakecase(data["selfWind"])] if data["selfWind"] is not None else None,
            round_wind=Wind[snakecase(data["roundWind"])] if data["roundWind"] is not None else None,
        )

File: mahjong-utils/src/commonMain/kotlin/mahjongutils/hora/HoraHandPattern.kt
@file:OptIn(ExperimentalSerializationApi::class)

package mahjongutils.hora

import kotlinx.serialization.EncodeDefault
import kotlinx.serialization.ExperimentalSerializationApi
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import mahjongutils.shanten.helpers.afterDiscardForHoraHand
import mahjongutils.models.*
import mahjongutils.models.hand.*

/**
 * 和牌手牌
 */
@Serializable
sealed interface HoraHandPattern : HoraInfo, HandPattern {
    companion object {
        /**
         * 构建和牌手牌
         *
         * @param pattern 手牌形
         * @param agari 和牌张
         * @param tsumo 是否自摸
         * @param selfWind 自风
         * @param roundWind 场风
         * @return 所有可能的和牌手牌
         */
        fun build(
            pattern: CommonHandPattern,
            agari: Tile,
            tsumo: Boolean,
            selfWind: Wind? = null,
            roundWind: Wind? = null
        ): Collection<HoraHandPattern> = when (pattern) {
            is RegularHandPattern -> RegularHoraHandPattern.build(pattern, agari, tsumo, selfWind, roundWind)
            is ChitoiHandPattern -> {
                require(pattern.pairs.size == 7 && pattern.remaining.isEmpty()) {
                    "this hand pattern is not hora yet"
                }
                listOf(ChitoiHoraHandPattern(pattern.pairs, agari, tsumo, selfWind, roundWind))
            }

            is KokushiHandPattern -> {
                require(pattern.yaochu == Tile.allYaochu && pattern.remaining.isEmpty() && pattern.repeated != null) {
                    "this hand pattern is not hora yet"
                }
                listOf(KokushiHoraHandPattern(pattern.repeated, agari, tsumo, selfWind, roundWind))
            }

            is HoraHandPattern -> listOf(pattern)
        }
    }
}

/**
 * 标准形的和牌手牌
 */
@Serializable
@SerialName("RegularHoraHandPattern")
data class RegularHoraHandPattern internal constructor(
    /**
     * 标准形手牌
     */
    val pattern: RegularHandPattern,
    override val agari: Tile,
    override val tsumo: Boolean,
    @EncodeDefault override val selfWind: Wind? = null,
    @EncodeDefault override val roundWind: Wind? = null,
    /**
     * 和牌搭子（为空表示单骑和牌）
     */
    val agariTatsu: Tatsu?,
) : HoraHandPattern, IRegularHandPattern by pattern {
    init {
        require(k == 4 && pattern.menzenMentsu.size + pattern.furo.size == 4 && pattern.jyantou != null && pattern.remaining.isEmpty()) {
            "this hand pattern is not hora yet"
        }
    }

    override val jyantou: Tile
        get() = pattern.jyantou!!

    companion object {
        private fun buildWithoutGot(
            agari: Tile,
            tsumo: Boolean,
            selfWind: Wind? = null,
            roundWind: Wind? = null,
            pattern: RegularHandPattern,
        ): RegularHoraHandPattern {
            if (pattern.jyantou != null) {
                return RegularHoraHandPattern(
                    agari = agari,
                    tsumo = tsumo,
                    selfWind = selfWind,
                    roundWind = roundWind,
                    agariTatsu = pattern.tatsu[0],
                    pattern = pattern.copy(
                        menzenMentsu = pattern.menzenMentsu + (pattern.tatsu[0].withWaiting(agari)),
                        tatsu = emptyList()
                    ),
                )
            } else {
                return RegularHoraHandPattern(
                    agari = agari,
                    tsumo = tsumo,
                    selfWind = selfWind,
                    roundWind = roundWind,
                    agariTatsu = null,
                    pattern = pattern.copy(
                        jyantou = agari,
                        remaining = emptyList()
                    ),
                )
            }
        }

        /**
         * 构建标准形和牌手牌
         *
         * @param pattern 手牌形
         * @param agari 和牌张
         * @param tsumo 是否自摸
         * @param selfWind 自风
         * @param roundWind 场风
         * @return 所有可能的和牌手牌
         */
        fun build(
            pattern: RegularHandPattern,
            agari: Tile,
            tsumo: Boolean,
            selfWind: Wind? = null,
            roundWind: Wind? = null
        ): Collection<RegularHoraHandPattern> {
            return pattern.afterDiscardForHoraHand(agari).map {
                buildWithoutGot(agari, tsumo, selfWind, roundWind, it)
            }
        }
    }
}

/**
 * 七对子的和牌手牌
 */
@Serializable
@SerialName("ChitoiHoraHandPattern")
data class ChitoiHoraHandPattern(
    /**
     * 手牌形
     */
    override val pairs: Set<Tile>,
    override val agari: Tile,
    override val tsumo: Boolean,
    @EncodeDefault override val selfWind: Wind? = null,
    @EncodeDefault override val roundWind: Wind? = null,
) : HoraHandPattern, IChitoiHandPattern {
    override val remaining: List<Tile>
        get() = emptyList()
}

/**
 * 国士无双的和牌手牌
 */
@Serializable
@SerialName("KokushiHoraHandPattern")
data class KokushiHoraHandPattern(
    /**
     * 手牌形
     */
    override val repeated: Tile,
    override val agari: Tile,
    override val tsumo: Boolean,
    @EncodeDefault override val selfWind: Wind? = null,
    @EncodeDefault override val roundWind: Wind? = null,
) : HoraHandPattern, IKokushiHandPattern {
    override val yaochu: Set<Tile>
        get() = Tile.allYaochu
    override val remaining: List<Tile>
        get() = emptyList()

    /**
     * 是否十三面和牌
     */
    val thirteenWaiting: Boolean
        get() = repeated == agari
}
